Add image and kernel builder

jcvenegas opened the pull request.

sameo commented on Makefile:
```diff
@@ -0,0 +1,54 @@
+		-v /dev:/dev \
+		$(RUN_PROXY) \
+		-ti \
+		-v $(CURDIR)/image_creator.sh:/image_creator.sh \
```
We could copy that one from the Dockerfile directly and make that command line a bit simpler.

jcvenegas replied:
done

sameo commented on Makefile:
```diff
@@ -0,0 +1,54 @@
+		-v $(CURDIR)/output:/output \
+		-v $(CURDIR)/rootfs:/content \
+		cc-image-builder \
+		bash -c "/image_creator.sh $(IMG_SIZE); cp container.img /output"
```
And this could be the default command (ENTRYPOINT) for the Dockerfile as well.

jcvenegas replied:
added an entrypoint

sameo commented on Makefile:
```diff
@@ -0,0 +1,54 @@
+		-v $(CURDIR)/build_cc_rootfs.sh:/build_cc_rootfs.sh \
+		$(RUN_PROXY) \
+		fedora:25 \
+		bash /build_cc_rootfs.sh
```
I would like this to be a little simpler:

* Have one single script that could either build a rootfs, a kernel, or a complete image depending on an option you pass it.
* Have one Dockerfile that uses that script as its ENTRYPOINT
* Call `docker run -d <rootfs|kernel|image>` on the built container

Basically, it would be better for the logic to be moved out of the Makefile, into the Dockerfile. Make the Makefile a `docker <build|run>` wrapper.

jcvenegas replied:
@sameo sure, thanks for the feedback.

jcvenegas replied:
@sameo 
1: Now  I there is an script that do almost all the logic, but not the image creation, take a look to the image_builder.sh script do you think we still need to added to the same script ? 
2: The new script  is ENTRYPOINT and have 3 options <rootfs|kernelimage>
3: The makefile now is a wrapper to build any of the options from osbuilder.sh

jcvenegas committed:
image: Add image_builder.sh

This patch adds the script image_builder.sh .
This scripts will create a Clear Containers image
based in a directory "rootfs" the current working
directory.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
add osbuilder.sh

This patch add osbuilder.sh, This scirpt will be used as
docker image entrypoint.

The script will take a parameter <kernel|rootfs|image>.

kernel: compiles a kernel from the "linux directory" in the current
working directory.

image: generate a container image using image_builder.sh

rootfs: create a rootfs based on clear linux packages.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
Add Dockerfile to build CC componets

This patch adds a new Dockerfile that uses
osbuilder.sh as entrypoint. This allow
to build CC components without install
extra dependence.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
Add Makefile docker run wrapper

Add a Makefile with the following targets:

make rootfs: Build a rootfs base for a CC image
make image: Build a CC image using rootfs directory
make kernel: Build a CC kernel based in linux directory

All the targets will work in the directory "workdir" in the
current working directory.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jodh-intel commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+	cd scripts; \
+	docker build $(BUILD_PROXY) -t $(IMAGE_BUILDER) . 
+
+#we only clone kernel if does not exit
```
typo: "exist"

jodh-intel commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+	make clear_containers_defconfig;
+
+workdir:
+	mkdir -p workdir
```
I can see that it's useful to have the directory below the source, but I don't think we can assume `$CWD` is writable.

jcvenegas replied:
agree, but in that case I dont know a good place to provide the image, kernel and rootfs where they will be modifying  and using.  I would not like to use a radom /tmp/directory. 

What about document it, and add variable in case the use would like to use a different directory?

jodh-intel replied:
Yes, being able to override a default value sounds like a good idea to me. In fact, you could add a `help` rule so the user can run `make help` and have that rule explain this - show the default value and explain how to override it.

jodh-intel commented on README.md:
```diff
@@ -1,2 +1,86 @@
+Clear Containers uses the Linux kernel, you can build a
+kernel compatible with Clear Containers using the make
+"kernel" target. This will clone the [Clear Container Kernel]
+if the directory "workidr/linux" does not exist, then will build it.
```
Typo: "workdir/linux".

jodh-intel commented on README.md:
````diff
@@ -1,2 +1,86 @@
+/usr/share/defaults/cc-oci-runtime/vm.json to /etc/cc-oci-runtime/vm.json
+
+```
+cp /usr/share/defaults/cc-oci-runtime/vm.json
````
Missing a destination? I'd delete line 56 and make this line:

```
sudo mkdir -p /etc/cc-oci-runtime
sudo cp /usr/share/defaults/cc-oci-runtime/vm.json /etc/cc-oci-runtime/vm.json
```

jodh-intel commented on README.md:
````diff
@@ -1,2 +1,86 @@
+cp /usr/share/defaults/cc-oci-runtime/vm.json
+```
+
+And modify the paths your new kernel and image:
````
"the paths for your"

jodh-intel commented on Makefile:
```diff
@@ -0,0 +1,63 @@
+#  Copyright (C) 2016 Intel Corporation
```
2017?

jodh-intel commented on scripts/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+From fedora:25
+
+RUN dnf install qemu-img parted gdisk make gcc bc -y
```
I'd put the `-y` before the list of packages as it's easier to see what is happening then.

jodh-intel commented on scripts/osbuilder.sh:
```diff
@@ -0,0 +1,85 @@
+build_rootfs()
+{
+    mkdir -p "${ROOTFS_DIR}"
+	DNF="dnf --config=$DNF_CONF -y --installroot=${ROOTFS_DIR} --noplugins"
```
Different indent here (and below) compared to line above.

jodh-intel commented on scripts/osbuilder.sh:
```diff
@@ -0,0 +1,85 @@
+build_kernel()
+{
+	pushd linux
+	make -j$(nproc)
```
You can drop `$(nproc)` and just call `make -j` which will then use all available CPUs by default.

grahamwhaley replied:
technically I believe if you just give `-j` with no parameters then `make` will (to quote from `man make`) *not limit the number of jobs that can run simultaneously*. It is not clear if that could then lead to a big thrash/swap if there were many more potentially parallel jobs than processors.

If you do go back to `$(nproc)`, then I'd suggest maybe `$(nproc)+1`, as that then (for some theories) always has one job waiting to launch when one of the running ones finishes. I'm not 100% convinced though. Your call I think?

jodh-intel commented on scripts/osbuilder.sh:
```diff
@@ -0,0 +1,85 @@
+	pushd linux
+	make -j$(nproc)
+	popd
+	cp linux/vmlinux vmlinuix.container
```
typo: "`vmlinux.container`".

jodh-intel commented on scripts/clear-dnf.conf:
```diff
@@ -0,0 +1,35 @@
+failovermethod=priority
+baseurl=https://download.clearlinux.org/current/x86_64/os/
+enabled=1
+gpgcheck=0
```
Ideally gpg signature checking would be enabled for `[clear]` and `[main]`.

If we cannot enable this, we really need comments here and a big scary warning in the `README.md` explaining why it is not enabled, and the potential associated risks.

jodh-intel commented on scripts/osbuilder.sh:
```diff
@@ -0,0 +1,85 @@
+Usage: ${script_name} rootfs|kernel|image
+rootfs : Build a rootfs based on Clear Linux packages
+kernel : Build a kernel for clear containers
+image  : Build a clear containers image based on rootfs directory
```
I'd capitalise ("Clear Containers") these 2 lines for consistency with "Clear Linux".

jodh-intel commented on scripts/image_builder.sh:
```diff
@@ -0,0 +1,60 @@
+
+# This scirpt creates a Clear Container image called container.img based
+# on the rootfs directory.
+
```
This is a short script, but I still think it's worth having a usage statement explaining what it does and how to call it.

jcvenegas replied:
added help information

jodh-intel commented on scripts/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+COPY osbuilder.sh         /usr/bin
+COPY image_builder.sh     /usr/bin
+COPY clear-dnf.conf       /etc/
```
Should this be copied to `/etc/dnf/`?

jodh-intel commented on scripts/clear-dnf.conf:
```diff
@@ -0,0 +1,35 @@
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+[main]
+cachedir=/var/cache/yum/clear/
```
Fedora 25 seems to use `/var/cache/dnf/` now (`yum` is fully deprecated).

jodh-intel commented on scripts/clear-dnf.conf:
```diff
@@ -0,0 +1,35 @@
+cachedir=/var/cache/yum/clear/
+keepcache=0
+debuglevel=2
+logfile=/var/log/yum.log
```
Again, looks like this should be `/var/log/dnf.log` for Fedora 25?

jodh-intel commented on scripts/clear-dnf.conf:
```diff
@@ -0,0 +1,35 @@
+gpgcheck=0
+plugins=0
+installonly_limit=3
+reposdir=/root/mash
```
Do we need this option?

jcvenegas replied:
Yes, it is only to avoid use fedora's defaults repos, added a note to explain that

jodh-intel commented on scripts/image_builder.sh:
```diff
@@ -0,0 +1,60 @@
+
+# Get the loop device bound to the image file (requires /dev mounted in the
+# image build system and root privileges)
+DEVICE=$(losetup -P -f --show ${IMAGE})
```
I think this script can only be run as `root`? If so, we need a check:

```
[ "$(id -u)" -eq 0 ] || die "$0: must be run as root"

```

jcvenegas replied:
done

jcvenegas commented:
@grahamwhaley  @jodh-intel  thanks a lot for your feedback , I have addressed your comments.

jodh-intel commented:
@jcvenegas - this is looking really good. There is just the question of the gpg signature checking.

jodh-intel commented on README.md:
```diff
@@ -1,2 +1,98 @@
+
+#### Clear Linux based packages security limitations  ####
+
+Clear Linux is not a based rpm Linux distribution and
```
"is not an rpm-based Linux distribution".

jodh-intel commented on README.md:
```diff
@@ -1,2 +1,98 @@
+#### Clear Linux based packages security limitations  ####
+
+Clear Linux is not a based rpm Linux distribution and
+the rpm packages are not signed, so there is not way
```
"there is no way"

jodh-intel commented on README.md:
```diff
@@ -1,2 +1,98 @@
+
+Clear Linux is not a based rpm Linux distribution and
+the rpm packages are not signed, so there is not way
+to ensure that downloaded packages not trustworthy.
```
"downloaded packages are trustworthy."

jodh-intel commented on README.md:
```diff
@@ -1,2 +1,98 @@
+
+If you are willing to use Clear Linux based images, we encourage you
+to use the Clear Containers images provided from its website
+https://download.clearlinux.org/current/ .
```
I don't think you need the space before the full-stop at the end of the sentence.

jodh-intel commented:
lgtm

jcvenegas commented:
@mcastelino @sameo do you have more feedback for this PR?

jodh-intel commented on Makefile:
```diff
@@ -0,0 +1,87 @@
+DOCKER_RUN=docker run \
+			--runtime runc \
+			--privileged \
+			-v /dev:/dev \
```
Nit: it would look cleaner to have both `-v` options next to each other.

sameo commented:
LGTM

mcastelino commented:
@jcvenegas @sameo LGTM. using this alongside https://github.com/jcvenegas/cc-oci-runtime/commit/3334a2bc7c915e1f631bc85355f1ce20ea984200 allows us to finally debug the rootfs.

@jcvenegas can you add documentation on
- how to add more packages to the rootfs (as the current rootfs has nothing besides the shell) which makes debug harder
- document how this can be used to debug the rootfs

I have not tried customizing the kernel yet with osbuilder, just rootfs

mcastelino commented on scripts/osbuilder.sh:
```diff
@@ -0,0 +1,85 @@
+{
+    mkdir -p "${ROOTFS_DIR}"
+    DNF="dnf --config=$DNF_CONF -y --installroot=${ROOTFS_DIR} --noplugins"
+	$DNF install systemd hyperstart cc-oci-runtime-extras coreutils systemd-bootchart
```
@jcvenegas may make sense to add an extra ENV variable that gets picked up here to add more packages

mcastelino commented on README.md:
````diff
@@ -1,2 +1,98 @@
+
+
+```
+sudo -E make kernel
````
@jcvenegas can we split this into two phases if possible. So that kernel customization is possible. i.e. modifying the config file. Just like we are doing today in the case of the rootfs and the image.

mcastelino commented:
@sameo can we merge this and improve on it over time?
